### DISCUSSION TITLE
H1 G2 Investigation into missing BMS Cell Temp (low/high) and Cell Volts (low/high) registers

### DISCUSSION THREAD

**Discussion by FozzieUK** [nathanmarlor/foxess_modbus#712] [Ideas]
Initial discussion started in Facebook with Richard W, now transferred to here 

> Hi Richard, here you go - sorry I did jot down the registers to test I just forgot to send it - i've also added the stop/start which I know works on the H1(G1) the G2 may well be the same as it's sort of a hybrid between the old spec and the new spec - see how you get on.
Check these addresses using the built in modbus read service (Developer Tools, Actions - select Foxess Modbus - Read Registers) selecxt your inverter in the drop down set the register address to read (try holding and input for all of these)
BMS1 MaxTemp		37617 I16 C  scale 10
BMS1 MinTemp		37618 I16 C  scale 10
BMS1 MaxCellVolts	37619 U16 V  scale 10
BMS1 MinCellVolts	37620 U16 V  scale 10
BMS1 SOH 		37624 U16 %  scale 1
BMS1 RemainEnergy	37632 U16 Wh scale 0.1
BMS1 FCC Capacity	37633 U16 Ah scale 10
BMS1 DesignEnergy	37635 U16 Wh scale 0.1
For Inverter Start/Stop
On the H1(G1) if you write to register 41013 with a value of 0 - it will stop the inverter, if you write to 41013 with a value of 1 it will start (after checking) - I know this works on my H1(G1).
The latest spec suggests that you now need to write to 49078 a value of 1 to shutdown the inverter, and if you write to 49077 a value of 1 it will power on the inverter.
(and if you read the value of 49228 it will show 1 if powered on, and 0 if powered off)
You can use the Foxess Modbus - Write Registers service for this.

**Comment by FozzieUK** [comment]
From Richard
>Thanks for these, looks like #37618 does give the lowest cell temp.  Results below.  There were no differences between input and holding (whatever the difference is?) and what is FCC Capacity? 
                                                                            input / holding
BMS1 MaxTemp		37617 I16 C  scale 10  - 241 / 241
BMS1 MinTemp		37618 I16 C  scale 10  - 189 / 189 (fox app 2 says 18.8 degrees, close enough 🙂 )
BMS1 MaxCellVolts	37619 U16 V  scale 10  - 33 / 33
BMS1 MinCellVolts	37620 U16 V  scale 10  - 33 / 33
BMS1 SOH 		37624 U16 %  scale 1       - 100 / 100
BMS1 RemainEnergy	37632 U16 Wh scale 0.1 - 1152 / 1152
BMS1 FCC Capacity	37633 U16 Ah scale 10  - 478 / 478
BMS1 DesignEnergy	37635 U16 Wh scale 0.1 - 1152 / 1152

**Reply by FozzieUK** [reply]
I think canton may be away at the moment - so over xmas I will work out what the changes might be, let you edit some of the entity definitions on your system to test them and hopefully we can get you a functional test system.

On the FCC capacity - not entirely sure as the spec doesn't provide detail, I think it might be the maximum charge capacity in Ah the battery is capable of when full (it looks like 47.8Ah for you ) so multiply that by the {maximum cell voltage= 3.6V} * {number of cells in each pack} should equal your individual battery pack Wh capability at full charge. But then equally that might be wrong - with some of these things we just have to track them and work out what they do over time and assess how to deal with them ☺️

The Cell low and high voltage is interesting as on the old specs they used to report in mV whereas now they report step 0.1 i.e. it was 3321 now 3.3V - this is a critical number to the BMS in how it calculates the pack SoC so i'm expecting this might change back in future firmware releases.

**Comment by u4h37u** [comment]
Thanks for looking into this @FozzieUK

**Reply by FozzieUK** [reply]
Hi Richard,

Here are some changes for you to run with and test, 

This will add the following sensors-

- inverter_fault_state
- battery voltage (as reported by BMS)
- battery current (as reported by BMS)
- bms_cell_temp_low
- bms_cell_temp_high
- bms_cell_mv_low
- bms_cell_mv_high

On the latest protocol spec (H3-Pro / G2 / KH) - the cell_mv_low / high are reported with the wrong scale as it is returning 33 with scale 10 (3.3V) - we used to get mV i.e. 3329mV - I think this could be an error as without that extra resolution these registers are not very effective, i'll leave it scaled wrong for now and we can change it later if Fox do correct it in later firmware releases.


For these changes i'm assuming you are on the latest release version of v1.12.2 for the line number references, but i've copied the before and after code so you can see precisely what's changing.


If you make the following changes to this file `/homeassistant/custom_components/foxess_modbus/entities/entity_descriptions.py`


These are changes - 


Line 1962 (bms_cell_temp_low/high, bms_cell_mv_low/high)


Was
```
        bms_cell_temp_high=[
            ModbusAddressesSpec(input=[11043], models=Inv.H1_G1 | Inv.KH_PRE119),
        ],
        bms_cell_temp_low=[
            ModbusAddressesSpec(input=[11044], models=Inv.H1_G1 | Inv.KH_PRE119),
        ],
        bms_cell_mv_high=[
            ModbusAddressesSpec(input=[11045], models=Inv.H1_G1 | Inv.KH_PRE119),
        ],
        bms_cell_mv_low=[
            ModbusAddressesSpec(input=[11046], models=Inv.H1_G1 | Inv.KH_PRE119),
```

becomes

```
        bms_cell_temp_high=[
            ModbusAddressesSpec(input=[11043], models=Inv.H1_G1 | Inv.KH_PRE119),
            ModbusAddressesSpec(holding=[37617], models=Inv.H1_G2),
        ],
        bms_cell_temp_low=[
            ModbusAddressesSpec(input=[11044], models=Inv.H1_G1 | Inv.KH_PRE119),
            ModbusAddressesSpec(holding=[37618], models=Inv.H1_G2),
        ],
        bms_cell_mv_high=[
            ModbusAddressesSpec(input=[11045], models=Inv.H1_G1 | Inv.KH_PRE119),
            ModbusAddressesSpec(holding=[37619], models=Inv.H1_G2),
        ],
        bms_cell_mv_low=[
            ModbusAddressesSpec(input=[11046], models=Inv.H1_G1 | Inv.KH_PRE119),
            ModbusAddressesSpec(holding=[37620], models=Inv.H1_G2),
```
---
Line 1945 (bat_volt and bat_current)

```
        batvolt=[
            ModbusAddressesSpec(input=[11034], models=Inv.H1_G1 | Inv.KH_PRE119),
            ModbusAddressesSpec(holding=[31034], models=Inv.H3_SET),
        ],
        bat_current=[
            ModbusAddressesSpec(input=[11035], models=Inv.H1_G1 | Inv.KH_PRE119),
            ModbusAddressesSpec(holding=[31035], models=Inv.H3_SET),
        ],
```

becomes

```
        batvolt=[
            ModbusAddressesSpec(input=[11034], models=Inv.H1_G1 | Inv.KH_PRE119),
            ModbusAddressesSpec(holding=[37609], models=Inv.H1_G2),
            ModbusAddressesSpec(holding=[31034], models=Inv.H3_SET),
        ],
        bat_current=[
            ModbusAddressesSpec(input=[11035], models=Inv.H1_G1 | Inv.KH_PRE119),
            ModbusAddressesSpec(holding=[37610], models=Inv.H1_G2),
            ModbusAddressesSpec(holding=[31035], models=Inv.H3_SET),
        ],
```

---
Line 1355 (inverter_fauult_code)

```
            ModbusAddressesSpec(holding=[39067, 39068, 39069], models=Inv.H3_PRO),
```

becomes

```
            ModbusAddressesSpec(holding=[39067, 39068, 39069], models=Inv.H3_PRO | Inv.H1_G2),
```

Then restart your HA check there are no errors at startup and have a look in devices & services, foxess_modbus to see if the additional sensors are showing data.

**Reply by u4h37u** [reply]
Hi Dave,

Thank you, much appreciated.  I've made the changes and restarted.  All are reporting sensor values which appear to be correct/expected values :-).    I'll monitor for a few days and report back.

Rich

**Reply by FozzieUK** [reply]
Ok great, let me know how it goes 👍  - hopefully you won’t need the fault state sensor but at least it’s operational now.

**Reply by u4h37u** [reply]
All have been reporting correctly since the modification, including the important cell temp low sensor:
![image](https://github.com/user-attachments/assets/7662ad4d-0545-4b86-8775-bca7003483f8)
As you say Cell mV High/Low are not reporting on the correct scale:
![image](https://github.com/user-attachments/assets/c35cd666-39a9-470d-9e5d-d7a36d692ef9)
I'll probably just disable those entities for now in HA.
Inverter fault code is useful and Battery Current has been handy too:
![image](https://github.com/user-attachments/assets/ce7490d0-d834-4001-b291-956be4e95296)

**Reply by FozzieUK** [reply]
Great, thanks - i’ve got a few things to do first and then i’ll get this written up into a PR 👍

**Reply by bfayers** [reply]
I made these changes on my install where my inverter is a "H1-6.0-E-G2" Manager: 1.37, Master: 1.22, Slave: 1.02

Had to remove the batvolt and bat_current definitions (37609,37610) because those fail to read with the illegal address error.

As for the other sensors, Inverter fault code is showing None (good, I assume), but BMS Cell Temp High/Low and BMS Cell mV High/Low all just report `0`. For what its worth the Fox app does show a much lower battery temperature than I see in HASS, so presumably that is the lowest cell temp

Possibly related -- the register for SoH in #703 confirmed as working on an H1 G2, also just returns 0 for me.

@icornish72 what versions do you have for manager/master/slave? I suspect I might need to ask for some firmware updates?

**Reply by FozzieUK** [reply]
@bfayers Ok thanks, it sounds like there are some firmware sensitivities, the movement of travel has been towards the new spec for a while - so at some point between your version and the latest it has added the bat (A/V) and BMS sensors.

@u4h37u Rich, could you confirm what your firmware versions are please

**Reply by bfayers** [reply]
> @bfayers Ok thanks, it sounds like there are some firmware sensitivities, the movement of travel has been towards the new spec for a while - so at some point between your version and the latest it has added the bat (A/V) and BMS sensors.

I'm thinking my version is pretty old -- going to request Fox to push updates, I've seen someone from all the way back in may last year get given 1.64 master 1.71 manager so I think mine must be on some pretty old builds.

**Reply by u4h37u** [reply]
I'm on Manager 1.44, Master 1.27, and Slave 1.02. 

I seem to recall some other improvements in this version over previous including ability to force charge and use PV at the same time.

**Reply by icornish72** [reply]
Master 1.27
Arm 1.44
Slave 1.02

As I said though, up until the last update, this was working fine.

**Reply by FozzieUK** [reply]
@u4h37u, @icornish72 thanks 👍 

> I'm thinking my version is pretty old -- going to request Fox to push updates, I've seen someone from all the way back in may last year get given 1.64 master 1.71 manager so I think mine must be on some pretty old builds.

The 1.64/1.71 is for the H1 (Gen 1).

Your firmware is definitely old - on the Fox cloud I can see that latest Manager(ARM) is 1.48 but Master is not currently showing so there may be a version being released (or withdrawn), but I would recommend getting it updated when you can and these extra registers will be available.

**Reply by mechanimal82** [reply]
I've just made these edits and restarted. Here's the values I'm seeing for each of the new sensors:
**inverter_fault_state:** None
**battery voltage (as reported by BMS):** 236.7 V
**battery current (as reported by BMS):** 0.0 A
**bms_cell_temp_low:** 0.6 °C
**bms_cell_temp_high:** 3.7 °C
**bms_cell_mv_low:** 3,287 mV
**bms_cell_mv_high:** 3,291 mV

Note that it is 22:00 at night, but the temperatures seem reasonable for the batteries currently in a garage with no insulation.

My system is: 
**Inverter:** H1-6.0-E-G2
**Version:** Manager: 1.48
**Version:** Master: 1.31
**Version:** Slave: 1.03

**Reply by u4h37u** [reply]
Ah interesting, you're seeing the correct values for BMS cell_mv_low and high.  I see you're on a slightly newer version of firmware than me which is probably it.

**Reply by FozzieUK** [reply]
Thanks for testing that, I had been hoping that Fox would get the bms_cell_mv_high/low scale corrected and that confirms it for Master 1.31, Manager 1.48

So we now know that for H1 G2:

Manager: 1.37, Master: 1.22 does not support all of the BMS registers (errors on batvolt and batcurrent)

Manager 1.44, Master 1.27 does support all the BMS registers but the scale is incorrect on bms_cell_mV-low/high

Manager 1.48, Master 1.31 does support all the BMS registers and the scale is correct on bms_cell_mV-low/high

NB it is the Manager (ARM) firmware that provides modbus and changes the modbus mapping.

@mechanimal82 would you mind doing a read testing for the SOH register (37624) when you get chance - (in HA Developer Tools, Actions - select Foxess Modbus - Read Registers) set register address 37624, count 1, holding and let me know what you see.

**Reply by bfayers** [reply]
I've now been updated to Manager 1.48, Master 1.31, Slave 1.02 (credit to Fox - it only took 1 email and to wait for the weekend to be over and it was done)

batvolt, and bat_current work as expected now, as do cell MV low/high and temp low/high. And the BMS kWh remaining sensor I mentioned in #702 works now too.

@FozzieUK while I'm here, I checked 37624 and I get "100", which is the same value I now see in the Fox app for SOH

**Reply by FozzieUK** [reply]
@bfayers that's great - thank you 👍

**Reply by mechanimal82** [reply]
> I've now been updated to Manager 1.48, Master 1.31, Slave 1.02 (credit to Fox - it only took 1 email and to wait for the weekend to be over and it was done)
> 
> batvolt, and bat_current work as expected now, as do cell MV low/high and temp low/high. And the BMS kWh remaining sensor I mentioned in #702 works now too.
> 
> @FozzieUK while I'm here, I checked 37624 and I get "100", which is the same value I now see in the Fox app for SOH

Yep, same for me. I have made the edits as mentioned in https://github.com/nathanmarlor/foxess_modbus/pull/703 and have the SOH showing as a sensor with a value of 100% within the Fox Modbus integration

Here's the output:
values:
  "37624": 100

**Reply by FozzieUK** [reply]
@mechanimal82  Perfect, thank you.

**Reply by KNZZZZZZZ** [reply]
@FozzieUK 
I also did these modifications on my HA Modbus v1.12.2 with the H1-G2 (Manager 1.48, Master 1.30, Slave 1.03).
I can see the new sensors but when I want to enable them I can't do it, it shows the message "Cannot change status. Disabled by Device."
How can I enable it?
![Screenshot](https://github.com/user-attachments/assets/ee88a178-dc48-4f69-9a0e-5108fb615c41)

Thanks!

**Reply by FozzieUK** [reply]
I’ve not seen that before with this integration, I think the easiest way to do this is to go to Settings, Devices & Services, click on the ‘>’ on the Foxess Modbus entry, then click on the 3 dots in the right hand corner - click ‘Disable’ say yes to the dialogue… wait for a few seconds and then click ENABLE
Hopefully that will work !

**Reply by KNZZZZZZZ** [reply]
Now it works indeed!
I suspect it had to do with my old H1-G1 inverter which is disabled in HA and had the same sensors. Thanks!!

**Comment by KNZZZZZZZ** [comment]
Register 37618 also gives the lowest cell temp on my H1-5.0-G2 inverter.
Is there a simple way to integrate this as a sensor?
Because I am still new to HA.

Thanks!

**Reply by FozzieUK** [reply]
If you don't mind editing the code on your system (under instruction), then keep an eye on this discussion as I will post some instructions on how to add this to the current build to test it - otherwise you'll have to wait a short while for all the tests to be completed, a request to be put in and a new version of the integration released.

**Comment by FozzieUK** [comment]
See https://github.com/nathanmarlor/foxess_modbus/pull/735

**Comment by FozzieUK** [comment]
@u4h37u , @mechanimal82

There is now a beta release v1.13.1b2 that includes these changes - the v1.13.x requires that HA core 2025.x is installed so when you get round to upgrading your ha core - could you please check this release and that everything is reporting as it should (including the SoH sensor).

**Reply by mechanimal82** [reply]
> @u4h37u , @mechanimal82
> 
> There is now a beta release v1.13.1b2 that includes these changes - the v1.13.x requires that HA core 2025.x is installed so when you get round to upgrading your ha core - could you please check this release and that everything is reporting as it should (including the SoH sensor).

Have installed B2 and rebooted. All sensors are populated, no errors or unknowns and the SOH, mV values etc are looking correct

**Reply by FozzieUK** [reply]
perfect, thanks 👍
